pkg/trie/trie.go

```diff
@@ -1,1518 +1,81 @@
 }
 
-func (t *Trie) buildEntriesMap(currentNode *Node, prefix []byte, kv map[string][]byte) {
-	if currentNode == nil {
-		return
-	}
-
-	// Leaf
-	if currentNode.Kind() == node.Leaf {
-		key := currentNode.PartialKey
-		fullKeyNibbles := concatenateSlices(prefix, key)
-		keyLE := codec.NibblesToKeyLE(fullKeyNibbles)
-		kv[string(keyLE)] = t.Get(keyLE)
-		return
-	}
-
-	// Branch
-	branch := currentNode
-	if branch.StorageValue != nil {
-		fullKeyNibbles := concatenateSlices(prefix, branch.PartialKey)
-		keyLE := codec.NibblesToKeyLE(fullKeyNibbles)
-		kv[string(keyLE)] = t.Get(keyLE)
-	}
-
-	for i, child := range branch.Children {
-		childPrefix := concatenateSlices(prefix, branch.PartialKey, intToByteSlice(i))
-		t.buildEntriesMap(child, childPrefix, kv)
-	}
+type DBBackedTrie interface {
+	Load(db db.DBGetter, rootHash common.Hash) error
+	WriteDirty(db NewBatcher) error
+	writeDirtyNode(db db.DBPutter, n *Node) (err error)
 }
 
-// NextKey returns the next key in the trie in lexicographic order.
-// It returns nil if no next key is found.
-func (t *Trie) NextKey(keyLE []byte) (nextKeyLE []byte) {
-	prefix := []byte(nil)
-	key := codec.KeyLEToNibbles(keyLE)
-
-	nextKey := findNextKey(t.root, prefix, key)
-	if nextKey == nil {
-		return nil
-	}
-
-	nextKeyLE = codec.NibblesToKeyLE(nextKey)
-	return nextKeyLE
+type Printable interface {
+	String() string
 }
 
-func findNextKey(parent *Node, prefix, searchKey []byte) (nextKey []byte) {
-	if parent == nil {
-		return nil
-	}
-
-	if parent.Kind() == node.Leaf {
-		return findNextKeyLeaf(parent, prefix, searchKey)
-	}
-	return findNextKeyBranch(parent, prefix, searchKey)
-}
-
-func findNextKeyLeaf(leaf *Node, prefix, searchKey []byte) (nextKey []byte) {
-	parentLeafKey := leaf.PartialKey
-	fullKey := concatenateSlices(prefix, parentLeafKey)
-
-	if keyIsLexicographicallyBigger(searchKey, fullKey) {
-		return nil
-	}
-
-	return fullKey
+type Versioned interface {
+	SetVersion(TrieLayout)
 }
 
-func findNextKeyBranch(parentBranch *Node, prefix, searchKey []byte) (nextKey []byte) {
-	fullKey := concatenateSlices(prefix, parentBranch.PartialKey)
-
-	if bytes.Equal(searchKey, fullKey) {
-		const startChildIndex = 0
-		return findNextKeyChild(parentBranch.Children, startChildIndex, fullKey, searchKey)
-	}
-
-	if keyIsLexicographicallyBigger(searchKey, fullKey) {
-		if len(searchKey) < len(fullKey) {
-			return nil
-		} else if len(searchKey) > len(fullKey) {
-			startChildIndex := searchKey[len(fullKey)]
-			return findNextKeyChild(parentBranch.Children,
-				startChildIndex, fullKey, searchKey)
-		}
-	}
-
-	// search key is smaller than full key
-	if parentBranch.StorageValue != nil {
-		return fullKey
-	}
-	const startChildIndex = 0
-	return findNextKeyChild(parentBranch.Children, startChildIndex,
-		fullKey, searchKey)
+type Hashable interface {
+	MustHash() common.Hash
+	Hash() (common.Hash, error)
+	GenesisBlock() (genesisHeader types.Header, err error)
```

Comment:
`GenesisBlock` does not seem to belong to this interface, why is the trie responsible for retrieving the genesis block? I think this responsibility is for the state manager who is aware of the chain

Reply:
Agree, but we are using it now in [node.go](https://github.com/ChainSafe/gossamer/blob/development/dot/node.go#L168) I think I could leave it for a different refactor PR